src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientBuilderExtensions.cs

```diff
@@ -645,16 +645,51 @@ public static IHttpClientBuilder ConfigureAdditionalHttpMessageHandlers(this IHt
             return builder;
         }
 
-#pragma warning disable 1591 // TODO: Document this API. https://github.com/dotnet/runtime/issues/105974
+        /// <summary>
+        /// Registers named <see cref="HttpClient"/> and the related handler pipeline <see cref="HttpMessageHandler"/> as keyed
+        /// services with the client's name as the key, and a lifetime provided in the <paramref name="lifetime" /> parameter.
+        /// By default, the lifetime is <see cref="ServiceLifetime.Scoped"/>.
+        /// </summary>
+        /// <param name="builder">The <see cref="IHttpClientBuilder"/>.</param>
+        /// <param name="lifetime">Lifetime of the keyed services registered.</param>
+        /// <returns>An <see cref="IHttpClientBuilder"/> that can be used to configure the client.</returns>
+        /// <remarks>
+        /// <para>
+        /// A named client resolved from DI as a keyed service will behave similarly to a client you would create with <see cref="IHttpClientFactory.CreateClient(string)"/>. This
+        /// means that the client will continue reusing the same <see cref="HttpMessageHandler"/> instance for the duration of <see cref="HttpClientFactoryOptions.HandlerLifetime"/>,
+        /// and it will continue to use the separate, handler's DI scope instead of the scope it was resolved from.
+        /// </para>
+        /// <para>
+        /// WARNING: Registering the client as a keyed <see cref="ServiceLifetime.Transient"/> service will lead to the <see cref="HttpClient"/> and <see cref="HttpMessageHandler"/>
+        /// instances being captured by DI as both implement <see cref="IDisposable"/>. This might lead to memory leaks if the client is resolved multiple times within a
+        /// <see cref="ServiceLifetime.Singleton"/> service.
```

Comment:
This seems pretty sad.

However, this isn't specific to the IHttpClientFactory, and the same issue exists with any disposable service, right? It's just a new foot gun for us since we couldn't be transient before?

Reply:
Yep 🥲 https://learn.microsoft.com/en-us/dotnet/core/extensions/dependency-injection-guidelines#disposable-transient-services-captured-by-container

But there is hope that this might be finally addressed (in some way) in 10.0 (this was discussed during API review, as it was one of the main blockers for keyed-by-default). Also see https://github.com/dotnet/runtime/issues/89755#issuecomment-2218758475. Tracking issue: https://github.com/dotnet/runtime/issues/36461 but we'd need to also create an additional one for per-service-descriptor opt-out.